src/pure_aworset.erl

```diff
@@ -0,0 +1,219 @@
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, query/1, equal/2]).
+-export([redundant/2, remove_redundant_crystal/2, remove_redundant_polog/2, check_stability/2]).
+
+-export_type([pure_aworset/0, pure_aworset_op/0]).
+
+-opaque pure_aworset() :: {?TYPE, payload()}.
+-type payload() :: {pure_type:polog(), ordsets:set()}.
+-type pure_aworset_op() :: {add, pure_type:element()} | {rmv, pure_type:element()}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new() -> pure_aworset().
+new() ->
+    {?TYPE, {orddict:new(), ordsets:new()}}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new([term()]) -> pure_aworset().
+new([]) ->
+    new().
+
+%% @doc Check redundancy `pure_aworset()'
+%% Called in remove_redundant().
+-spec redundant({pure_type:id(), pure_aworset_op()}, {pure_type:id(), pure_aworset_op()}) ->
+    integer().
+redundant({VV1, {add, Elem1}}, {VV2, {_X, Elem2}}) ->
+    case Elem1 =:= Elem2 andalso pure_trcb:happened_before(VV1, VV2) of
+        true ->
+            1;
+        false ->
+            0
+    end.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_polog({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_polog({VV1, Op}, {?TYPE, {POLog0, ORSet}}) ->
+    POLog1 = orddict:fold(
+        fun(Key, Value, Acc) ->
+            case redundant({Key, Value}, {VV1, Op}) of
+                0 ->
+                    orddict:store(Key, Value, Acc);
+                1 ->
+                    Acc
+            end
+        end,
+        orddict:new(),
+        POLog0
+    ),
+    {true, {?TYPE, {POLog1, ORSet}}}.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_crystal({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_crystal({_VV1, {_X, Elem}}, {?TYPE, {POLog, AWORSet}}) ->
+    case ordsets:is_element(Elem, AWORSet) of
+        true ->
+            {true, {?TYPE, {POLog, ordsets:del_element(Elem, AWORSet)}}};
+        false ->
+            {true, {?TYPE, {POLog, AWORSet}}}
+    end.
+
+%% @doc Checks stable operations and remove them from POLog of `pure_aworset()'
+-spec check_stability(pure_type:id(), pure_aworset()) -> pure_aworset().
+check_stability(StableVV, {?TYPE, {POLog0, AWORSet0}}) ->
+    {POLog1, AWORSet1} = orddict:fold(
+        fun(Key, {_Op, Elem}=Value, {AccPOLog, AccORSet}) ->
+            case pure_trcb:happened_before(Key, StableVV) of
+                true ->
+                    {AccPOLog, ordsets:add_element(Elem, AccORSet)};
+                false ->
+                    {orddict:store(Key, Value, AccPOLog), AccORSet}
+            end
+        end,
+        {orddict:new(), AWORSet0},
+        POLog0
+    ),
+    {?TYPE, {POLog1, AWORSet1}}.
```

Comment:
This stability check is linear on the size of the Polog as a new POLog1 is always constructed. Maybe one should document that this code assumes a small Polog, so that it is changed in the future if dealing with large ones.